### PR TITLE
Jm mb 1250 zip2rate models

### PR DESCRIPTION
**Pull request by jacquelineIO**
## Description

Go model creation for `re_zip5_rate_areas` and  and updates for table `re_zip3s`. 

Follow-on work to add an additional constraint will be done under story: https://dp3.atlassian.net/browse/MB-480
```
alter table re_zip3s
    constraint re_zip3s_has_zip5 CHECK (
	(has_multiple_rate_areas = true and rate_area_id is null) or (has_multiple_rate_areas = false and rate_area_id is not null)
	);
```

## Reviewer Notes

n/a

## Setup

Add any steps or code to run in this section to help others prepare to run your code:

```sh
make server_test
```

## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](https://github.com/transcom/mymove/tree/master/docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](https://github.com/transcom/mymove/tree/master/docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](https://github.com/transcom/mymove/tree/master/docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
  * [ ] Secure migrations have been tested using `scripts/run-prod-migrations`
* [ ] Request review from a member of a different team.
* [ ] Have the Jira acceptance criteria been met for this change?

## References

* [Jira story](https://dp3.atlassian.net/browse/MB-1250) for this change
* [Zip-to-Rate Area Design Proposal](https://docs.google.com/document/d/1qCD5Z6qtPgasqySVtZPTceb9xoGt4KHk_XLwP36dDOc/edit#) Design story.

## Screenshots

n/a

### CODE REVIEW

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+    rate_area_id uuid not null
+        constraint re_zip5_rate_areas_rate_area_id_fkey references re_rate_areas,
+    zip5 text not null
+        constraint re_zip5_rate_areas_zip5_key unique,
+    created_at timestamp not null,
+    updated_at timestamp not null
+);
+
+alter table re_zip3s
+    add rate_area_id uuid null
```
Comment (lynzt) [inline migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: we prob also want an index on `rate_area_id`

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql [hunk 2]
```diff
@@ -0,0 +1,16 @@
+create table re_zip5_rate_areas
```
Comment (lynzt) [inline migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: assume we'd want indexes on at least `rate_area_id`, and prob `zip5` if we're going to use `zip5` in a where clause...

Comment (jacquelineIO) [reply migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: done

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql (hunk 1 above)
Comment (jacquelineIO) [reply migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: done

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql (hunk 2 above)
Comment (chrisgilmerproj) [reply migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: If you are constantly querying them together you can also make a combined index. Which means that they'll be indexed together for the best efficiency. If you're going to query only one or the other then what you have is fine. I'd simply add:

```
CREATE INDEX re_zip5_rate_areas_id_to_zip5 ON re_zip5_rate_areas (rate_area_id, zip5);
```

And also below add 

```
CREATE INDEX re_zip3_rate_areas_id_to_zip3 ON re_zip3s (rate_area_id, zip3);
```

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql (hunk 1 above)
Comment (reggieriser) [reply migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: If `zip5` is already marked as `unique`, I believe it should already have an index.  I'm not sure if explicitly adding a separate one means we now have two indexes on the same column.

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    rate_area_id uuid not null
+        constraint re_zip5_rate_areas_rate_area_id_fkey references re_rate_areas,
+    zip5 text not null
+        constraint re_zip5_rate_areas_zip5_key unique,
+    created_at timestamp not null,
+    updated_at timestamp not null
+);
+
+CREATE INDEX ON re_zip5_rate_areas (rate_area_id);
+CREATE INDEX ON re_zip5_rate_areas (zip5);
```
Comment (reggieriser) [inline migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: If `zip5` is already marked as `unique`, I believe it should already have an index.  I'm not sure if explicitly adding a separate one means we now have two indexes on the same column.

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+alter table re_zip3s
+    add rate_area_id uuid null
+     constraint re_zip3s_rate_area_id_fkey references re_rate_areas,
+    add has_multiple_rate_areas boolean not null default false;
+
+CREATE INDEX ON re_zip3s (rate_area_id);
+CREATE INDEX ON re_zip3s (zip3);
```
Comment (reggieriser) [inline migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: Same thing with `zip3` here ... do we need a separate index if it's already a `unique` column?

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql (hunk 3 above)
Comment (reggieriser) [reply migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: From the [Postgres docs](https://www.postgresql.org/docs/10/indexes-unique.html) ... "There's no need to manually create indexes on unique columns; doing so would just duplicate the automatically-created index."

File: migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql (hunk 2 above)
Comment (reggieriser) [reply migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: Just as an addendum to what @chrisgilmerproj said ... generally, a combined index of `(rate_area_id, zip5)` should work for querying both together **or** querying `rate_area_id` on its own.  However, a query on just `zip5` won't be able to take full advantage of that index ([reference](https://www.postgresql.org/docs/10/indexes-multicolumn.html)).

Comment (jacquelineIO) [reply migrations/app/schema/20200214175204_create_re_zip5_rate_areas_table.up.sql]: > Multicolumn indexes should be used sparingly. In most situations, an index on a single column is sufficient and saves space and time. 

Sounds like we have to be certain how we are using the data. The code isn't written right now, so I'll leave it as is.
thank you!

**Pull request merged**